x-pack/auditbeat/module/system/user/users_linux.go

```diff
@@ -150,24 +167,33 @@ func readShadowFile() (map[string]shadowFileEntry, error) {
 		shadow := shadowFileEntry{
 			// sp_lstchg is in days since Jan 1, 1970.
 			LastChanged: epoch.AddDate(0, 0, int(spwd.sp_lstchg)),
-		}
 
-		/*
-			Make sure the full password hash is long (it should be) and then collect only 10 characters.
-			This should be long enough to make collisions (where a new password hash shares these
-			10 characters with an old password hash) extremely unlikely. In addition, we also check
-			the day the password was last changed to detect a password change.
-
-			The hash excerpt is not included in any event output, but it is persisted to the
-			beat.db file on disk.
-		*/
-		passwordHash := C.GoString(spwd.sp_pwdp)
-		if len(passwordHash) > 24 {
-			shadow.PasswordHashExcerpt = passwordHash[len(passwordHash)-10:]
+			// The password hash is never output to Elasticsearch or any other output,
+			// but a hash of the hash is persisted to disk in the beat.db file.
+			Password: C.GoString(spwd.sp_pwdp),
 		}
 
 		shadowEntries[C.GoString(spwd.sp_namp)] = shadow
 	}
 
 	return shadowEntries, nil
 }
+
+// noSaltHash takes a password string from /etc/passwd or /etc/shadow, tries to skip the salt,
+// and returns a SHA-512 hash of the rest.
+func noSaltHash(password string) []byte {
+	// On modern UNIX systems, the format of the password field is $id$salt$encrypted (see crypt(3)).
+	// We only want to use the `encrypted` part.
```

Comment:
Why do we only want the encrypted part?

Reply:
I was aiming to reduce the amount of sensitive data we hold on to (even in hashed form). The encrypted part is the hash that should be unique on its own and so should be good enough for our purpose of detecting a change.

Reply:
Given that you are hashing the the entire value I don't see any reason to filter out the salt. Plus, IIUC, by filtering it out your lose the ability to detect changes to the salt. The seems like small amount of logic that could be dropped.

Reply:
> Plus, IIUC, by filtering it out your lose the ability to detect changes to the salt.

I don't think the salt can change without the hash changing, too - given it's a part of the input to the hashing function?

I agree that filtering out the salt provides debatable benefits: An attacker would have to gain access to `beat.db`, successfully reverse the SHA-512 hash, and then compute a rainbow table for that specific salt value. But if we don't need the salt to detect a change - why take even a theoretical risk?

Reply:
If the salt is changed in the file without changing the password hash the user will not be able to log in. At login the machine will compare the stored password hash with the output of `hash(salt+user_provided_password)`. Changing only the salt is not something that should happen, but I'd like to know if any part of the value changed (salt, algorithm, or hash).

> successfully reverse the SHA-512 hash

Unlike encryption, hashing is not reversible so this isn't a possibility.

Reply:
> > successfully reverse the SHA-512 hash
> 
> Unlike encryption, hashing is not reversible so this isn't a possibility.

Is it never? I assume most collision attacks find inputs that produce the same hash, but are not actually the same input as the original that produced the hash in the first place (i.e. `hash(m1) == hash(m2)` but `m1 != m2`). But is that always the case, or could `m1 == m2`?

Reply:
For posterity: We talked about this and agreed that there is little risk in storing a SHA-512 hash of the password field in a local database file that's only accessible by `root` (same as `/etc/shadow`).